### PR TITLE
many: replace state.State restart support with overlord/restart

### PR DESCRIPTION
**Pull request by pedronis**
this offers functionality equivalent to state.State.RequestRestart
etc but on top of the state itself instead of tied into it

a small difference is that this required the state lock being held
consistently while some subset of the old functionality didn't need
it

### CODE REVIEW

File: overlord/restart/restart_test.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+	defer st.Unlock()
+
+	// unitialized
+	ok, t := restart.Pending(st)
+	c.Check(ok, Equals, false)
+	c.Check(t, Equals, restart.RestartUnset)
+
+	h := &testHandler{}
+
+	err := restart.Init(st, "boot-id-1", h)
```
Comment (mvo5) [inline overlord/restart/restart_test.go:70]: (nitpick^2) I suppose we could test that errors from the handler are returned as expected but given that we test that the handler is caleld it probably does not add enough for the extra work.

File: overlord/hookstate/hookmgr.go [hunk 2]
```diff
@@ -336,13 +350,9 @@ func (m *HookManager) runHook(context *Context, snapst *snapstate.SnapState, hoo
 
 	if hookExists || mustHijack {
 		// we will run something, not a noop
-		if ok, _ := m.state.Restarting(); ok {
-			// don't start running a hook if we are restarting
-			return &state.Retry{}
+		if err := m.runGuard(context); err != nil {
```
Comment (mvo5) [inline overlord/hookstate/hookmgr.go:353]: (nitpick) maybe a comment here that `runGuard` checks for restarting or something? The name is quite generic so a bit more context might be helpful for the causal reader :) (but I tried to come up with a more expressive name for the func and couldn't so maybe it's fine).

File: overlord/snapstate/handlers_discard_test.go [hunk 3]
```diff
@@ -38,7 +38,7 @@ type discardSnapSuite struct {
 var _ = Suite(&discardSnapSuite{})
 
 func (s *discardSnapSuite) SetUpTest(c *C) {
-	s.setup(c, nil)
```
Comment (mvo5) [inline overlord/snapstate/handlers_discard_test.go:41]: Thanks for this cleanup

File: overlord/snapstate/handlers_link_test.go [hunk 4]
```diff
@@ -49,36 +50,27 @@ import (
...
-
-func (b *witnessRestartReqStateBackend) EnsureBefore(time.Duration) {}
-
 func (s *linkSnapSuite) SetUpTest(c *C) {
-	s.stateBackend = &witnessRestartReqStateBackend{}
+	s.baseHandlerSuite.SetUpTest(c)
 
-	s.setup(c, s.stateBackend)
+	s.state.Lock()
+	restart.Init(s.state, "boot-id-0", snapstatetest.MockRestartHandler(func(t restart.RestartType) {
```
Comment (mvo5) [inline overlord/snapstate/handlers_link_test.go:62]: This is nice

File: overlord/restart/restart.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package restart implements requesting restarts from any part of the code that has access to state.
+package restart
+
+import (
+	/*"encoding/json"
```
Comment (mvo5) [inline overlord/restart/restart.go:24]: (stating the obvious) this can go before we merge I suppose :)

**Review by mvo5 — COMMENTED** [review]
Thank you, this looks very nice. Some small comments inline - what is missing before this can go from draft->real-PR? it looks fairly good to me already.

**Comment by pedronis** [comment]
> Some small comments inline - what is missing before this can go from draft->real-PR? it looks fairly good to me already.

@mvo5  mostly marked it draft because it's a bit big and could be split in 2 or 3 PRs if preferred

File: overlord/restart/restart.go (hunk 5 above)
Comment (pedronis) [reply overlord/restart/restart.go:24]: oops, sorry, I forgot about this

File: overlord/hookstate/hookmgr.go (hunk 2 above)
Comment (pedronis) [reply overlord/hookstate/hookmgr.go:353]: runHookRestartingGuard ?

**Review by mvo5 — APPROVED** [review]
Thank you

**Comment by mvo5** [comment]
I think it's fine as is, I reviewed the individual commits which make it fairly straightforward IMHO - but I'm not opposed to splitting too.

File: overlord/restart/restart_test.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+	return nil
+}
+
+func (s *restartSuite) TestRequestRestartDaemon(c *C) {
+	st := state.New(nil)
+
+	st.Lock()
+	defer st.Unlock()
+
+	// unitialized
```
Comment (MiguelPires) [inline overlord/restart/restart_test.go:63]: Nitpick
```suggestion
	// uninitialized
```

**Review by MiguelPires — APPROVED** [review]
LGTM

**Comment by pedronis** [comment]
marking blocked as #10871 should land first now

**Pull request merged**